cpp/src/copying/gather.cuh

```diff
@@ -0,0 +1,337 @@
+		     gather_map+num_destination_rows, source_data,
+		     destination_data);
+    }
+
+    CHECK_STREAM(stream);
+    return destination_column;
+  }
+
+  template <typename column_type, typename iterator_type,
+    std::enable_if_t<not is_fixed_width<column_type>()>* = nullptr>
+  std::unique_ptr<column> operator()(column_view const& source_column,
+				     iterator_type gather_map,
+				     size_type num_destination_rows,
+				     bool ignore_out_of_bounds,
+				     util::cuda::scoped_stream stream) {
+    CUDF_FAIL("Column type must be numeric");
+  }
+
+};
+
+/**
+ * @brief Specifies the behavior of index_converter
+ */
+enum index_conversion {
+    NEGATIVE_TO_POSITIVE = 0,
+    NONE
+};
+
+/**---------------------------------------------------------------------------*
+* @brief Function object for applying a transformation on the gathermap
+* that converts negative indices to positive indices
+*
+* A negative index `i` is transformed to `i + size`, where `size` is
+* the number of elements in the column being gathered from.
+* Allowable values for the index `i` are in the range `[-size, size)`.
+* Thus, when gathering from a column of size `10`, the index `-1`
+* is transformed to `9` (i.e., the last element), `-2` is transformed
+* to `8` (the second-to-last element) and so on.
+* Positive indices are unchanged by this transformation.
+*---------------------------------------------------------------------------**/
+template <typename map_type, index_conversion ic=index_conversion::NONE>
+struct index_converter : public thrust::unary_function<map_type,map_type>{};
+
+
+template <typename map_type>
+struct index_converter<map_type, index_conversion::NEGATIVE_TO_POSITIVE>
+{
+  index_converter(size_type n_rows)
+  : n_rows(n_rows) {}
+
+  __device__
+  map_type operator()(map_type in) const
+  {
+    return ((in % n_rows) + n_rows) % n_rows;
+  }
+  size_type n_rows;
+};
+
+template <typename map_type>
+struct index_converter<map_type, index_conversion::NONE>
+{
+  index_converter(size_type n_rows)
+  : n_rows(n_rows) {}
+
+  __device__
+  map_type operator()(map_type in) const
+  {
+    return in;
+  }
+  size_type n_rows;
+};
+
+
+template <typename iterator_type>
+std::unique_ptr<table>
+gather(table_view const& source_table, iterator_type gather_map,
+       size_type num_destination_rows, bool check_bounds = false,
+       bool ignore_out_of_bounds = false,
+       bool allow_negative_indices = false,
+       rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource())
+{
+  auto source_n_cols = source_table.num_columns();
+  auto source_n_rows = source_table.num_rows();
+
+  std::vector<util::cuda::scoped_stream> v_stream(source_n_cols);
+
+  std::vector<std::unique_ptr<column>> destination_columns;
+
+  for (size_type i = 0; i < source_n_cols; i++) {
+    column_view src_col = source_table.column(i);
+    // Perform sanity checks
+    CUDF_EXPECTS(src_col.data<void*>() != nullptr, "Missing source data buffer");
```

Comment:
This shouldn't be possible anymore given the requirements of `table_view`/`column_view` constructors.